src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/VideoLinkAppointmentLinker.kt

```diff
@@ -0,0 +1,116 @@
+import uk.gov.justice.digital.hmpps.whereabouts.model.VideoLinkBooking
+import uk.gov.justice.digital.hmpps.whereabouts.repository.VideoLinkAppointmentRepository
+import uk.gov.justice.digital.hmpps.whereabouts.repository.VideoLinkBookingRepository
+import javax.transaction.Transactional
+
+typealias AppointmentFinder = (PrisonAppointment?) -> VideoLinkAppointment?
+
+@Service
+class VideoLinkAppointmentLinker(
+  private val videoLinkAppointmentRepository: VideoLinkAppointmentRepository,
+  private val videoLinkBookingRepository: VideoLinkBookingRepository,
+  private val prisonApiService: PrisonApiService
+) {
+
+  fun linkAppointments() {
+
+    getBookingIdsOfUnlinkedAppointments().forEach { linkAppointmentsForOffenderBookingId(it) }
+  }
+
+  @Transactional
+  fun getBookingIdsOfUnlinkedAppointments(): List<Long> {
+    val bookingIds = videoLinkAppointmentRepository.bookingIdsOfUnlinkedAppointments()
+    log.info("Found unlinked VideoLinkAppointments for ${bookingIds.size} bookingIds")
+    return bookingIds
+  }
+
+  @Transactional
+  fun linkAppointmentsForOffenderBookingId(bookingId: Long) {
+    val bookings = videoLinkBookingsForOffenderBookingId(bookingId)
+
+    log.info("bookingId $bookingId: Creating ${bookings.size} VideoLinkBookings")
+
+    videoLinkBookingRepository.saveAll(bookings)
+  }
+
+  fun videoLinkBookingsForOffenderBookingId(bookingId: Long): List<VideoLinkBooking> {
+    val prisonAppointments = getPrisonVideoLinkAppointmentsForBookingId(bookingId)
+    val prisonAppointmentsById = prisonAppointments.associateBy { it.eventId }
+
+    val mainAppointments = videoLinkAppointmentRepository.unlinkedMainAppointmentsForBookingId(bookingId)
+
+    log.info("bookingId $bookingId: Found ${mainAppointments.size} MAIN VideoLinkAppointments")
+
+    val findPreAppointmentFor = preAppointmentFinder(bookingId, prisonAppointments)
+    val findPostAppointmentFor = postAppointmentFinder(bookingId, prisonAppointments)
+
+    return mainAppointments
+      .filter { prisonAppointmentsById.containsKey(it.appointmentId) }
+      .map { main ->
+        val mainPrisonAppointment = prisonAppointmentsById[main.appointmentId]
+
+        VideoLinkBooking(
+          main = main,
+          pre = findPreAppointmentFor(mainPrisonAppointment),
+          post = findPostAppointmentFor(mainPrisonAppointment),
+        )
+      }
+  }
+
+  fun getPrisonVideoLinkAppointmentsForBookingId(bookingId: Long): List<PrisonAppointment> {
+    val prisonAppointments = mutableListOf<PrisonAppointment>()
+    var offset = 0
+
+    do {
+      val chunk = prisonApiService
+        .getPrisonAppointmentsForBookingId(bookingId, offset, CHUNK_SIZE)
+        .filter { it.eventSubType == "VLB" }
+      prisonAppointments.addAll(chunk)
+      offset += chunk.size
+    } while (chunk.isNotEmpty())
+
+    log.info("bookingId $bookingId: Retrieved ${prisonAppointments.size} VLB appointments from prison-api")
+
+    return prisonAppointments
+  }
+
```

Comment:
`preAppointmentFinder` and `postAppointmentFinder` return (a) PRE/POST appointment contiguous with the supplied appointment. For pre appointments this means `pre.endTime == main.startTime` and for post `post.startTime == main.endTime`.  This may be too strict. 
If so then these two functions are where a better appointment linking algorithm should be defined.